### PR TITLE
Factor out an `is_mine_server_name` method

### PR DESCRIPTION
**Pull request by squahtx**
Add an `is_mine_server_name` method, similar to `is_mine_id`.

Ideally we would use this consistently, instead of sometimes comparing
against `hs.hostname` and other times reaching into
`hs.config.server.server_name`.

Also fix a bug in the tests where `hs.hostname` would sometimes differ
from `hs.config.server.server_name`.

### CODE REVIEW

File: tests/unittest.py [hunk 1]
```diff
@@ -585,15 +587,25 @@ def setup_test_homeserver(self, *args: Any, **kwargs: Any) -> HomeServer:
...
+
         # Parse the config from a config dict into a HomeServerConfig
         config_obj = make_homeserver_config_obj(config)
         kwargs["config"] = config_obj
 
+        # The server name in the config is now `name`, if provided, or the `server_name`
+        # from a config override, or the default of "test". Whichever it is, we
+        # construct a homeserver with a matching name.
+        kwargs["name"] = config_obj.server.server_name
+
```
Comment (squahtx) [inline tests/unittest.py:603]: We're inconsistent about whether we use `hs.hostname` or `hs.config.server.server_name` in Synapse. Prior to this change, they could contain entirely different values during the tests.

Comment (clokep) [reply tests/unittest.py:603]: Having to shove this *back* into `kwargs` is a bit annoying, but oh well.

**Review by clokep — APPROVED** [review]

File: tests/unittest.py (hunk 1 above)
Comment (squahtx) [reply tests/unittest.py:603]: We could also copy this back into `name` and pass `name` to `setup_test_homeserver`. It's unpleasant either way.

File: synapse/federation/federation_base.py [hunk 2]
```diff
@@ -50,6 +50,7 @@ def __init__(self, hs: "HomeServer"):
         self.hs = hs
 
         self.server_name = hs.hostname
```
[the remark this answers is not in the document]
Comment (squahtx) [reply synapse/federation/federation_base.py]: Sounds reasonable, let's move it there.

File: tests/unittest.py (hunk 1 above)
Comment (clokep) [reply tests/unittest.py:603]: Yeah, all the config code around tests is annoying. I think this is fine.

**Pull request merged**